File: wiki_java/src/main/java/com/github/klee0kai/wiki/wrapping/GoodPhone.java
package com.github.klee0kai.wiki.wrapping;

import com.github.klee0kai.stone.Stone;
import com.github.klee0kai.stone.wrappers.LazyProvide;
import com.github.klee0kai.stone.wrappers.PhantomProvide;
import com.github.klee0kai.test.tech.components.Battery;
import com.github.klee0kai.test.tech.components.Ram;

import javax.inject.Inject;

public class GoodPhone {

    static TechFactoryComponent DI = Stone.createComponent(TechFactoryComponent.class);

    @Inject
    public PhantomProvide<Battery> battery;

    @Inject
    public LazyProvide<Ram> ram;

    public void create() {
        DI.inject(this);
    }

}

File: wiki_kotlin/src/main/kotlin/com/github/klee0kai/wiki/wrapping/GoodPhone.kt
package com.github.klee0kai.wiki.wrapping

import com.github.klee0kai.stone.Provider
import com.github.klee0kai.stone.wrappers.LazyProvider
import com.github.klee0kai.test.tech.components.Battery
import com.github.klee0kai.test.tech.components.Ram
import javax.inject.Inject

val DI: TechFactoryComponent = TechFactoryComponentStoneComponent()

class GoodPhone {

    @Inject
    lateinit var battery: Provider<Battery>

    @Inject
    lateinit var ram: LazyProvider<Ram>

    fun create() {
        DI.inject(this)
    }

}
